### PR TITLE
Added JSONDecodeError for rapidjson >= 0.7.1 with fallback to ValueError.

### LINKED ISSUE #172: Bug in JSON error checking
### Description

The rapidjson package does not have a `decoder` submodule. Since we don't have a corresponding compatibility layer, things like `json.decoder.JSONDecodeError` fail.

### To reproduce

```
$ signac init json_decoder
$ signac find '{"x": True}'
27958648a9e57fcd66ae5e31ff3359e9
$ signac find '{"x": True}'  # This is invalid JSON, the expected behavior is to raise an error indicating that the JSON is invalid
Error: module 'signac.core.json' has no attribute 'decoder'  # The actual output
```

### PR DESCRIPTION
**Pull request by bdice**
## Motivation and Context
Resolves #172.

## Types of Changes
- [x] Bug fix

## Checklist:
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

### CODE REVIEW

File: signac/core/json.py [hunk 1]
```diff
@@ -16,6 +16,11 @@
 try:
     import rapidjson as json
     from rapidjson import Encoder
+    try:
+        # Defined for rapidjson >= 0.7.1
+        from rapidjson import JSONDecodeError
+    except ImportError:
+        JSONDecodeError = ValueError
```
Comment (bdice) [inline signac/core/json.py]: Before 0.7.1, the error thrown was a ValueError. This should ensure that the error in `contrib.filterparse` still gets caught.

**Comment by bdice** [comment]
I'd like to add a step to CI that runs tests with rapidjson if we're committed to supporting it as a primary JSON backend, going forward.

**Comment by vyasr** [comment]
fwiw I think when we last discussed this we came down on not knowing whether rapidjson was really the right JSON backend to choose, but we needed to do a more comprehensive survey of libraries to know what to support and we hadn't assigned that to anyone.

**Comment by vyasr** [comment]
I'm not sure I follow exactly what additional test you want to add, but this PR looks good to me right now. One minor suggestion, for redirecting stderr you might want to just copy over the functions we use in testing signac-flow.

**Comment by bdice** [comment]
> fwiw I think when we last discussed this we came down on not knowing whether rapidjson was really the right JSON backend to choose, but we needed to do a more comprehensive survey of libraries to know what to support and we hadn't assigned that to anyone.

I did some reading on that subject while working on this PR. Seeing [this comment on rapidjson's issue tracker](https://github.com/python-rapidjson/python-rapidjson/issues/111#issuecomment-424706000) suggests to me that we would be better served by `ujson` or `simplejson` or another alternative to `rapidjson` that provides direct support for the standard library's `json` API.

**Comment by bdice** [comment]
> I'm not sure I follow exactly what additional test you want to add, but this PR looks good to me right now. One minor suggestion, for redirecting stderr you might want to just copy over the functions we use in testing signac-flow.

@vyasr I added the test already, so that's done. I also made the change you suggested to use signac-flow's redirection.

I think this PR is ready to merge.

**Comment by vyasr** [comment]
I agree, should be ready for merge.

And regarding choice of library, to my knowledge ujson provides the best balance of compatibility with the Python library's json module and performance, but I haven't done exhaustive research into that. We might just need to make a quick test where replace the code in signac's json module with the relevant import and see how performance improves/degrades. Actually changing which modules we support should not be too hard once we know the correct solution that provides the closest mirror to the existing functionality we expect.

**Comment by vyasr** [comment]
I just went to look at issues and found #171 which I made :D good to know we're rehashing and coming to the same conclusions, it looks like your benchmarks agree with ujson as the best alternative. @csadorf if you want to assign that issue to a particular benchmark we can assign it to someone to work on.

**Pull request merged**